pkg/render/csr.go

```diff
@@ -0,0 +1,122 @@
+func CreateCSRInitContainer(
+	installation *operator.InstallationSpec,
+	cm *operator.CertificateManagement,
+	mountName string,
+	commonName string,
+	keyName string,
+	certName string,
+	dnsNames []string,
+	registerApiserver bool) corev1.Container {
+	return corev1.Container{
+		Name:            "key-cert-provisioner",
+		Image:           components.GetReference(components.ComponentCSRInitContainer, "docker.io/", installation.ImagePath), //todo: Set the right registry.
+		ImagePullPolicy: "Always",                                                                                            //todo: delete this line.
+		VolumeMounts: []corev1.VolumeMount{
+			{MountPath: "/secret", Name: mountName, ReadOnly: false},
+		},
+		Env: []corev1.EnvVar{
+			{Name: "SECRET_LOCATION", Value: "/secret/"},
+			{Name: "SIGNER", Value: cm.SignerName},
+			{Name: "COMMON_NAME", Value: commonName},
+			{Name: "KEY_ALGORITHM", Value: fmt.Sprintf("%v", installation.CertificateManagement.KeyAlgorithm)},
+			{Name: "SIGNATURE_ALGORITHM", Value: fmt.Sprintf("%v", installation.CertificateManagement.SignatureAlgorithm)},
+			{Name: "REGISTER_APISERVER", Value: fmt.Sprintf("%v", registerApiserver)},
+			{Name: "KEY_NAME", Value: keyName},
+			{Name: "CERT_NAME", Value: certName},
+			{Name: "DNS_NAMES", Value: strings.Join(dnsNames, ",")},
+			{Name: "POD_IP", ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "status.podIP",
+				},
+			}},
+			{Name: "POD_NAME", ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.name",
+				},
+			}},
+			{Name: "POD_NAMESPACE", ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.namespace",
+				},
+			}},
+		},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged:               Bool(false),
+			AllowPrivilegeEscalation: Bool(false),
+		},
+	}
+}
+
+// csrClusterRole returns a role with the necessary permissions to create certificate signing requests.
+func csrClusterRole() runtime.Object {
+	return &rbacv1.ClusterRole{
+		TypeMeta: metav1.TypeMeta{Kind: "ClusterRole", APIVersion: "rbac.authorization.k8s.io/v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "tigera-csr-creator",
+			Labels: map[string]string{},
+		},
+
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{"certificates.k8s.io"},
+				Resources: []string{"certificatesigningrequests"},
+				Verbs:     []string{"create", "watch"},
+			},
+		},
+	}
+}
+
+// csrClusterRoleBinding returns a role binding with the necessary permissions to create certificate signing requests.
+func csrClusterRoleBinding(name, namespace string) *rbacv1.ClusterRoleBinding {
+	crb := &rbacv1.ClusterRoleBinding{
+		TypeMeta: metav1.TypeMeta{Kind: "ClusterRoleBinding", APIVersion: "rbac.authorization.k8s.io/v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   fmt.Sprintf("%s-%s:csr-creator", namespace, name),
```

Comment:
Why do we need the namespace since this is a cluster role ?

Reply:
I suppose it is not really necessary, I'll remove it.